packages/backend/src/app.ts

```diff
@@ -622,43 +632,44 @@ export class App {
         },
         validatorMiddlewareOptions
       ),
+      getWalletAddressUrlFromQuote,
       createTokenIntrospectionMiddleware({
         requestType: AccessType.Quote,
         requestAction: RequestAction.Read
       }),
       httpsigMiddleware,
+      getWalletAddressForSubresource,
       quoteRoutes.get
     )
 
     router.get(
       WALLET_ADDRESS_PATH + '/jwks.json',
-      createWalletAddressMiddleware(),
-      createValidatorMiddleware<WalletAddressKeysContext>(
+      createValidatorMiddleware(
         walletAddressServerSpec,
         {
           path: '/jwks.json',
           method: HttpMethod.GET
         },
         validatorMiddlewareOptions
       ),
-      async (ctx: WalletAddressKeysContext): Promise<void> =>
-        await walletAddressKeyRoutes.getKeysByWalletAddressId(ctx)
+      getWalletAddressUrlFromPath,
+      walletAddressKeyRoutes.get
     )
 
     // Add the wallet address query route last.
     // Otherwise it will be matched instead of other Open Payments endpoints.
     router.get(
       WALLET_ADDRESS_PATH,
-      createWalletAddressMiddleware(),
-      createSpspMiddleware(this.config.spspEnabled),
-      createValidatorMiddleware<WalletAddressContext>(
+      createValidatorMiddleware(
         walletAddressServerSpec,
         {
           path: '/',
           method: HttpMethod.GET
         },
         validatorMiddlewareOptions
       ),
+      getWalletAddressUrlFromPath,
+      createSpspMiddleware(this.config.spspEnabled),
```

Comment:
do not call `getWalletAddressForSubresource` here and in the jwks.json route since we have different behaviours in these routes for handling `deps.config.walletAddressUrl === ctx.walletAddressUrl`